cyvlfeat/fisher/tests/fisher_test.py

```diff
@@ -7,48 +7,73 @@ def test_fisher_dimension():
-    means = np.random.uniform(size=(D, K)).astype(np.float32)
-    covariances = np.random.uniform(size=(D, K)).astype(np.float32)
+
+    x = np.random.uniform(size=(N, D)).astype(np.float32)
+    means = np.random.uniform(size=(K, D)).astype(np.float32)
+    covariances = np.random.uniform(size=(K, D)).astype(np.float32)
     priors = np.random.uniform(size=(K,)).astype(np.float32)
     priors /= np.linalg.norm(priors)
     enc = fisher(x, means, covariances, priors, verbose=True)
-    
+
     expected = 2 * K * D
     observed = len(enc)
     assert(expected == observed)
 
 
+def _compute_posteriors(diff, covars, prior):
+    q = (diff ** 2).sum(-1)  # Mahalanobis distances to cluster centers
+    # Compute the log probability of samples knowing each cluster
+    q = -0.5 * (q + np.log(covars).sum(-1) +
+                diff.shape[1] * np.log(2 * np.pi))
+    # Compute the joint probability of xi and j-th cluster
+    # p(xi, cj) = p(xi|cj) * p(cj)
+    q += np.log(prior)
+    # Normalize values to avoid numerical problems
+    qmax = q.max(-1)
+    q -= qmax[:, None]
+    # Compute the posteriors
+    q = np.exp(q)
+    # normalize
+    q /= q.sum(-1)[:, np.newaxis]
+    return q
+
+
+def _fisher_encoding(X, means, covars, priors):
+    diff = (X[:, None] - means[None]) / np.sqrt(covars)
+    q = _compute_posteriors(diff, covars, priors)
+    u = (diff * q[..., np.newaxis]).mean(0) / np.sqrt(priors)[:, np.newaxis]
+    v = ((diff ** 2 - 1) * q[..., np.newaxis]).mean(0) / \
+        np.sqrt(2 * priors)[:, np.newaxis]
+    return np.hstack([u.ravel(), v.ravel()])
+
+
 def test_fisher_encoding():
-    N = 21
-    K = 3
-    D = 5
-    x = np.zeros(D * N, dtype=np.float32)
-    for i in range(D * N):
-        x[i] = i
-    x = x.reshape(D, N)
-
-    mu = np.zeros(D * K, dtype=np.float32)
-    for i in range(D * K):
-        mu[i] = i
-    mu = mu.reshape(D, K)
-    
-    sigma2 = np.ones((D, K), dtype=np.float32)
-    prior = (1.0 / K) * np.ones(K, dtype=np.float32)
-
-    observed_enc = fisher(x, mu, sigma2, prior, verbose=True)
-
-    # expected result obtained from running vl_fisher_encode from a C program
-    expected_enc = np.array([0.000000000, 0.000000000, 0.000000000,
-                             0.000000000, 0.000000000, 0.000000000,
-                             0.000000000, 0.000000000, 0.000000000,
-                             0.000000000, 70.519210815, 70.519210815,
-                             70.519210815, 70.519210815, 70.519210815,
-                             -0.058321182, -0.058321182, -0.058321182,
-                             -0.058321182, -0.058321182, -0.058321182,
-                             -0.058321182, -0.058321182, -0.058321182,
-                             -0.058321182, 3073.876220703, 3073.876220703,
-                             3073.876220703, 3073.876220703, 3073.876220703],
-                            dtype=np.float32)
-    assert_allclose(expected_enc, observed_enc)
```

Comment:
Would it be possible to maintain this test and just check that the input matrix transpose is still equal to this?